Reader: Follow conversation should handle `success: false`

mindgraffiti opened the pull request:
Ref. #15642 

In this PR
- the taskSucceeded bool continues to bubble up the success blocks until it reaches one where user feedback is handled
- if the user fails to follow the conversation, a toast message appears with a generic error message and reverts the Follow conversation button

**Not implemented:** better error messaging to the user. Also not implemented is a nudge for users to check their Reader subscription notification settings.  

### To test:
1. Log in with a non-a8c WordPress account, with a validated email address. 
2. On the web, disable email notifications by visiting [Reader subscription notification settings](https://wordpress.com/me/notifications/subscriptions) and checking the box in the "Block emails" section.
2. Go to Reader > Discover and find a blog post that has the comment button enabled
3. Tap the comment button, then tap the "Follow conversation by email" button
4. A toast message will appear, "Unable to subscribe to the comments" and the follow conversation button will revert from "**Following** conversation by email" to "**Follow** conversation..."

(A user will always be allowed to unfollow a conversation.)

PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

mindgraffiti committed:
Handle the bool passed into the success block

mindgraffiti committed:
Continue handling the taskSucceeded bool from previous success blocks

and add some basic user messaging when success is false.

mindgraffiti committed:
Force the podfile to pick up the beta version of WordPressKit

mindgraffiti committed:
Merge branch 'develop' into issue/15642-handle-success-false

mindgraffiti committed:
Update podfile.lock after merge conflicts were resolved

mindgraffiti committed:
Merge branch 'develop' into issue/15642-handle-success-false

mindgraffiti committed:
Update unit tests to account for new param in success block

mindgraffiti committed:
Merge remote-tracking branch 'origin/issue/15642-handle-success-false' into issue/15642-handle-success-false

ScoutHarris commented on WordPress/Classes/ViewRelated/Reader/ReaderCommentsViewController.m:
```diff
@@ -1215,15 +1215,28 @@ - (void)handleFollowConversationButtonTapped
+    void (^successBlock)(BOOL taskSucceeded) = ^void(BOOL taskSucceeded) {
+        if (taskSucceeded == NO) {
+            NSString *title = newIsSubscribed
+                ? NSLocalizedString(@"Unable to subscribe to the comments", @"The app failed to subscribe to the comments for the post")
```
I wonder if these messages should be more in line with the button text, and other toast messages. Maybe like `Unable to follow conversation`, `Successfully followed conversation`, etc?

mindgraffiti committed:
Merge branch 'develop' into issue/15642-handle-success-false

mindgraffiti committed:
Try to clarify the success / fail messages to the user

This is not its final form

mindgraffiti commented on WordPress/Classes/ViewRelated/Reader/ReaderCommentsViewController.m:
```diff
@@ -1215,15 +1215,28 @@ - (void)handleFollowConversationButtonTapped
+    void (^successBlock)(BOOL taskSucceeded) = ^void(BOOL taskSucceeded) {
+        if (taskSucceeded == NO) {
+            NSString *title = newIsSubscribed
+                ? NSLocalizedString(@"Unable to follow conversation", @"The app failed to subscribe to the comments for the post")
+                : NSLocalizedString(@"Failed to unfollow conversation", @"The app failed to unsubscribe from the comments for the post");
```
@ScoutHarris this is not the final form of the error handling or the error messaging. I will be adding a second PR for that.

ScoutHarris replied:
Ah gotcha!

ScoutHarris reviewed (changes requested):
Hey @mindgraffiti .

The button state isn't updated correctly when unfollowing a conversation.

- Uncheck 'block emails'.
- Go to Reader > Discover and find a blog post that has the comment button enabled.
- Tap the comment button, then tap the "Follow conversation by email" button.
- See toast message.
- Tap the "Following conversation by email" button" to unfollow.
- I get the unfollowed toast message, but the button shows 'Follow' for a second then changes back to 'Following'.


https://user-images.githubusercontent.com/1816888/109050177-5816e700-7696-11eb-90b2-9efe7c45c551.mp4

mindgraffiti committed:
Change UI in the main thread; update the button only if task fails

Because we already optimistically change the button state, we only change the button again if the task failed or the request failed.

mindgraffiti commented:
Nice catch. As soon as I re-read my success block I saw the error 🤦‍♀️ . Thank you for testing more than the described states!

ScoutHarris reviewed (approved):
LGTM!

:shipit:
